### PR TITLE
Update id_set to support packs

### PR DESCRIPTION
**Pull request by Shellyber**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/19876

## Description
Updated the id_set_update file to support content packs.

## Must have
- [x] Tests
- [ ] Code Review

### CODE REVIEW

File: Tests/scripts/update_id_set.py [hunk 1]
```diff
@@ -305,12 +309,11 @@ def process_integration(file_path):
     """
     res = []
     if os.path.isfile(file_path):
-        if re.match(INTEGRATION_REGEX, file_path, re.IGNORECASE) or \
-                re.match(BETA_INTEGRATION_REGEX, file_path, re.IGNORECASE):
+        if checked_type(file_path, (INTEGRATION_REGEX, BETA_INTEGRATION_REGEX, PACKS_INTEGRATION_YML_REGEX)):
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]:
```suggestion
        if checked_type(file_path, (INTEGRATION_REGEX, BETA_INTEGRATION_REGEX, PACKS_INTEGRATION_REGEX)):
```
if it doesn't exist, add it.

types should consist of the following:
- isfile
* regular integration
* pack integration
* beta integration
- else
* regular package integration
* pack package integration
* beta package integration

File: Tests/scripts/update_id_set.py [hunk 2]
```diff
@@ -305,12 +309,11 @@ def process_integration(file_path):
...
             print("adding {0} to id_set".format(file_path))
             res.append(get_integration_data(file_path))
-    else:  # In case we encountered a package
-        for yml_file in glob.glob(os.path.join(file_path, '*.yml')):
+    else:
+        for yml_file in glob.glob(os.path.join(file_path, os.path.basename(file_path) + '.yml')):
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: no need to iterate, remove the for.
to be safe, add `os.path.isfile` to check that the file actually exists.

File: Tests/scripts/update_id_set.py [hunk 3]
```diff
@@ -348,13 +358,66 @@ def process_testplaybook_path(file_path):
...
         playbook = get_playbook_data(file_path)
+
     return playbook, script
 
 
+def get_integrations_paths():
+    path_list = [
+        ['Integrations', '*'],
+        ['Beta_Integrations', '*'],
+        ['Packs', '*', 'Integrations', '*', '*.yml'],  # Case of package
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: why are you considering packages only for packs and not integrations/beta?
no need to split, the `process_integration` will handle that

File: Tests/scripts/update_id_set.py [hunk 4]
```diff
@@ -348,13 +358,66 @@ def process_testplaybook_path(file_path):
...
+    for path in path_list:
+        integration_files.extend(glob.glob(os.path.join(*path)))
+
+    return integration_files
+
+
+def get_scripts_paths():
+    path_list = [
+        ['Scripts', '*'],
+        ['Packs', '*', 'Scripts', '*', '*.yml'],  # Case of package
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: same here

File: Tests/scripts/update_id_set.py [hunk 5]
```diff
@@ -348,13 +358,66 @@ def process_testplaybook_path(file_path):
...
+    for path in path_list:
+        script_files.extend(glob.glob(os.path.join(*path)))
+
+    return script_files
+
+
+def get_playbooks_paths():
+    path_list = [
+        ['Playbooks', '*.yml'],
+        ['Packs', '*', 'Playbooks', '*.yml']
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: we also have playbooks in the beta folder

File: Tests/scripts/update_id_set.py [hunk 6]
```diff
@@ -396,6 +460,76 @@ def re_create_id_set():
     exec_time = time.time() - start_time
     print_color("Finished the creation of the id_set. Total time: {} seconds".format(exec_time), LOG_COLORS.GREEN)
 
+    duplicates = find_duplicates(new_ids_dict)
+    if duplicates:
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: `duplicates = [[], [], [], []]` is a `true` expression but indicates that there are no duplicates.
```suggestion
    if any(duplicates):
```

File: Tests/scripts/update_id_set.py [hunk 7]
```diff
@@ -396,6 +460,76 @@ def re_create_id_set():
...
+        is_duplicate = True
+        if dict1_from_version >= dict2_to_version:
+            is_duplicate = False
+        if dict2_to_version > dict1_to_version:
+            is_duplicate = False
+        if dict2_from_version > dict1_to_version:
+            is_duplicate = False
+        if dict1_from_version > dict2_from_version:
+            is_duplicate = False
+    return is_duplicate
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: when you encounter a problem, you should exit immediately - no reason to continue the iterations.
look at my original comment (previous PR):
if (A, C) is the first pair, the other pairs will override the value of `is_duplicate` and the function will return false.

File: Tests/scripts/update_id_set.py [hunk 8]
```diff
@@ -121,7 +125,7 @@ def get_integration_data(file_path):
     cmd_list = [command.get('name') for command in commands]
 
     integration_data['name'] = name
-    integration_data['file_path'] = file_path
+    integration_data['path'] = file_path
```
[the remark this answers is not in the document]
Comment (Shellyber) [reply Tests/scripts/update_id_set.py]: didn't appear first at my original file. 
Should I change it back or leave it as it is?

[the remark this answers is not in the document]
Comment (yardensade) [reply Tests/scripts/update_id_set.py]: master says file_path, not sure why is it changed.

[the remark this answers is not in the document]
Comment (Shellyber) [reply Tests/scripts/update_id_set.py]: fine

File: Tests/scripts/update_id_set.py [hunk 9]
```diff
@@ -1,26 +1,30 @@
...
 from Tests.scripts.constants import *  # noqa: E402
 from Tests.test_utils import get_yaml, get_to_version, get_from_version, collect_ids, get_script_or_integration_id, \
-    LOG_COLORS, print_color, run_command  # noqa: E402
+    LOG_COLORS, print_color, run_command, print_error  # noqa: E402
 
 
 CHECKED_TYPES_REGEXES = (INTEGRATION_REGEX, PLAYBOOK_REGEX, SCRIPT_REGEX,
-                         TEST_PLAYBOOK_REGEX, INTEGRATION_YML_REGEX)
+                         TEST_PLAYBOOK_REGEX, INTEGRATION_YML_REGEX, PACKS_INTEGRATION_YML_REGEX,
+                         PACKS_SCRIPT_YML_REGEX, PACKS_PLAYBOOK_YML_REGEX, PACKS_TEST_PLAYBOOKS_REGEX)
```
[the remark this answers is not in the document]
Comment (Shellyber) [reply Tests/scripts/update_id_set.py]: Why I need the missing ones? I'm handling all the yml files

**Comment by Itay4** [comment]
whats the status here?

File: Tests/scripts/update_id_set.py [hunk 10]
```diff
@@ -348,13 +371,65 @@ def process_testplaybook_path(file_path):
     print("adding {0} to id_set".format(file_path))
     script = None
     playbook = None
-    if re.match(TEST_SCRIPT_REGEX, file_path, re.IGNORECASE):
+    if checked_type(file_path, (TEST_SCRIPT_REGEX, PACKS_TEST_PLAYBOOKS_REGEX)):
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: or whatever the correct name is...
```suggestion
    if checked_type(file_path, (TEST_SCRIPT_REGEX, PACKS_TEST_SCRIPT_REGEX)):
```

File: Tests/scripts/update_id_set.py [hunk 11]
```diff
@@ -305,12 +321,12 @@ def process_integration(file_path):
     """
     res = []
     if os.path.isfile(file_path):
-        if re.match(INTEGRATION_REGEX, file_path, re.IGNORECASE) or \
-                re.match(BETA_INTEGRATION_REGEX, file_path, re.IGNORECASE):
+        if checked_type(file_path, (INTEGRATION_REGEX, BETA_INTEGRATION_REGEX, PACKS_INTEGRATION_YML_REGEX,
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: `PACKS_INTEGRATION_YML_REGEX` should not be here. it should be in the else part as it relates to the integration package (dir)

File: Tests/scripts/update_id_set.py [hunk 12]
```diff
@@ -396,6 +472,73 @@ def re_create_id_set():
...
+    for dup1, dup2 in itertools.combinations(duplicates, 2):
+        dict1 = dup1.values()[0]
+        dict2 = dup2.values()[0]
+        dict1_from_version = LooseVersion(dict1.get('fromversion', '0.0.0'))
+        dict2_from_version = LooseVersion(dict2.get('fromversion', '0.0.0'))
+        dict1_to_version = LooseVersion(dict1.get('toversion', '99.99.99'))
+        dict2_to_version = LooseVersion(dict2.get('toversion', '99.99.99'))
+
+        if dict1['name'] != dict2['name']:
+            print_error('The following objects has the same ID but different names: {}, {}.'.format(dict1['name'],
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]:
```suggestion
            print_warning('The following objects has the same ID but different names: "{}", "{}".'.format(dict1['name'],
```

File: Tests/scripts/update_id_set.py (hunk 2 above)
Comment (yaakovi) [reply Tests/scripts/update_id_set.py]: you missed this one

File: Tests/scripts/update_id_set.py [hunk 13]
```diff
@@ -396,6 +472,73 @@ def re_create_id_set():
     exec_time = time.time() - start_time
     print_color("Finished the creation of the id_set. Total time: {} seconds".format(exec_time), LOG_COLORS.GREEN)
 
+    duplicates = find_duplicates(new_ids_dict)
+    if any(duplicates):
+        print_error('The following duplicates were found: {}'.format(duplicates))
+        sys.exit(1)
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: this will fail master, must be remove right before merging.

File: Tests/scripts/update_id_set.py [hunk 14]
```diff
@@ -396,6 +472,73 @@ def re_create_id_set():
     exec_time = time.time() - start_time
     print_color("Finished the creation of the id_set. Total time: {} seconds".format(exec_time), LOG_COLORS.GREEN)
 
+    duplicates = find_duplicates(new_ids_dict)
```
Comment (yaakovi) [inline Tests/scripts/update_id_set.py]: performance wise - no need to run this every build, should be removed to validate_files in the future (and verified only on relevant IDs and not all).

**Comment by Shellyber** [comment]
@yaakovi @yardensade @Itay4 How do we differ playbook and script in test playbook folder without the prefix?

File: Tests/scripts/update_id_set.py (hunk 14 above)
Comment (yardensade) [reply Tests/scripts/update_id_set.py]: I'm facing the issue of which files to test in secrets too, to be honest I don't think it should happen inside the scripts but only in the file validator. With a circle/no circle distinction

File: Tests/scripts/infrastructure_tests/update_id_set_test.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+
+class TestIntegration(unittest.TestCase):
+    def test_get_integration_data(self):
+        """
+        Test for getting all the integration data
+        """
+        file_path = 'Packs/CortexXDR/Integrations/PaloAltoNetworks_XDR/PaloAltoNetworks_XDR.yml'
+        # if not os.path.isfile(file_path):
```
Comment (yardensade) [inline Tests/scripts/infrastructure_tests/update_id_set_test.py]: may we remove these comments?

File: Tests/scripts/update_id_set.py [hunk 16]
```diff
@@ -219,7 +219,7 @@ def get_script_data(file_path, script_code=None):
 
 def get_depends_on(data_dict):
     depends_on = data_dict.get('dependson', {}).get('must', [])
-    depends_on_list = list(set([cmd.split('|')[-1] for cmd in depends_on]))
+    depends_on_list = list(set(cmd.split('|')[-1] for cmd in depends_on))
```
Comment (yardensade) [inline Tests/scripts/update_id_set.py]: Shouldn't we split on 3 |||| ?

Comment (yaakovi) [reply Tests/scripts/update_id_set.py]: actually yes, did not looked at the logic of the code, just fixed the lint suggestion.
either way it works the same

**Pull request merged**